From 1b36a0266447e44b21ba69ab4da3c1e6382afda2 Mon Sep 17 00:00:00 2001
From: Gabriel Przybysz Gonçalves Júnior <gabriel.przybsyzgoncalvesjunior@windriver.com>
Date: Mon, 17 Mar 2025 14:22:00 -0300
Subject: [PATCH] csi-powerstore: removing leases from k8s

This patch consists of removing leases created on application-apply
command.

Signed-off-by: Gabriel Przybysz Gonçalves Júnior <gabriel.przybsyzgoncalvesjunior@windriver.com>
---
 .../templates/job-post-delete.yaml            | 132 ++++++++++++++++++
 charts/csi-powerstore/values.yaml             |   3 +
 2 files changed, 135 insertions(+)
 create mode 100644 charts/csi-powerstore/templates/job-post-delete.yaml

diff --git a/charts/csi-powerstore/templates/job-post-delete.yaml b/charts/csi-powerstore/templates/job-post-delete.yaml
new file mode 100644
index 0000000..2d63b32
--- /dev/null
+++ b/charts/csi-powerstore/templates/job-post-delete.yaml
@@ -0,0 +1,132 @@
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Release.Name }}-clear-sa
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+    "helm.sh/hook": "post-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation, hook-succeeded"
+    "helm.sh/hook-weight": "-5"
+imagePullSecrets:
+  - name: default-registry-key
+
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ .Release.Name }}-clear-cr
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+    "helm.sh/hook": "post-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation, hook-succeeded"
+    "helm.sh/hook-weight": "-5"
+rules:
+  - apiGroups: ["coordination.k8s.io"]
+    resources: ["leases"]
+    verbs: ["list", "delete"]
+
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ .Release.Name }}-clear-crb
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+    "helm.sh/hook": "post-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation, hook-succeeded"
+    "helm.sh/hook-weight": "-5"
+subjects:
+  - kind: ServiceAccount
+    name: {{ .Release.Name }}-clear-sa
+    namespace: {{ .Release.Namespace }}
+roleRef:
+  kind: ClusterRole
+  name: {{ .Release.Name }}-clear-cr
+  apiGroup: rbac.authorization.k8s.io
+
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-clear-bin
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+    "helm.sh/hook": "post-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation, hook-succeeded"
+    "helm.sh/hook-weight": "-5"
+data:
+  clear_script.sh: |-
+    #!/bin/bash
+
+    echo "Checking for leases in ${NAMESPACE} namespace..."
+
+    if ! kubectl get leases -n ${NAMESPACE} --no-headers | grep -q .; then
+      echo "No leases found. Exiting."
+      exit 0
+    fi
+
+    # Clear dell-storage leases
+    echo "Removing leases from ${NAMESPACE} namespace"
+    kubectl delete leases -n ${NAMESPACE} --all
+
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ .Release.Name }}-clear-job
+  namespace: {{ .Release.Namespace }}
+  labels:
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    chart: "{{.Chart.Name}}"
+    app.starlingx.io/component: {{ ternary "application" "platform" .Values.labels.isApplication }}
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+    "helm.sh/hook": "post-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation, hook-succeeded"
+spec:
+  backoffLimit: 5  # Limit the number of job restart in case of failure: ~5 minutes.
+  template:
+    metadata:
+      name: {{ .Release.Name }}-clear-job
+      namespace: {{ .Release.Namespace }}
+      labels:
+        heritage: {{ .Release.Service | quote }}
+        release: {{ .Release.Name | quote }}
+        chart: "{{.Chart.Name}}"
+        app.starlingx.io/component: {{ ternary "application" "platform" .Values.labels.isApplication }}
+    spec:
+      serviceAccountName: {{ .Release.Name }}-clear-sa
+      nodeSelector: { "node-role.kubernetes.io/control-plane": "" }
+      tolerations:
+      - effect: NoSchedule
+        operator: Exists
+        key: node-role.kubernetes.io/master
+      - effect: NoSchedule
+        operator: Exists
+        key: node-role.kubernetes.io/control-plane
+      restartPolicy: OnFailure
+      volumes:
+        - name: {{ .Release.Name }}-clear-bin
+          configMap:
+            name: {{ .Release.Name }}-clear-bin
+      containers:
+        - name: clear-script
+          image: {{ .Values.kubectl.image | quote }}
+          command: [ "/bin/bash", "/tmp/mount/clear_script.sh" ]
+          env:
+            - name: DRIVER_NAME
+              value: {{ .Values.driverName }}
+            - name: NAMESPACE
+              value: {{ .Release.Namespace }}
+          volumeMounts:
+            - name: {{ .Release.Name }}-clear-bin
+              mountPath: /tmp/mount/
diff --git a/charts/csi-powerstore/values.yaml b/charts/csi-powerstore/values.yaml
index ad863ee..6c59d6e 100644
--- a/charts/csi-powerstore/values.yaml
+++ b/charts/csi-powerstore/values.yaml
@@ -37,6 +37,9 @@ version: v2.13.0
 driver:
   image: quay.io/dell/container-storage-modules/csi-powerstore:v2.13.0

+kubectl:
+  image: docker.io/bitnami/kubectl:1.29
+
 # Specify kubelet config dir path.
 # Ensure that the config.yaml file is present at this path.
 # Default value: /var/lib/kubelet
--
2.34.1

